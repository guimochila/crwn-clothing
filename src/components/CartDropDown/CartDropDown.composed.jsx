import React from 'react';
import { withRouter } from 'react-router-dom';
import { Query, Mutation } from 'react-apollo';
import { adopt } from 'react-adopt';

import CartDropDown from './CartDropDown';
import { TOGGLE_CART_HIDDEN, GET_CART_ITEMS } from '../../graphql/queries';

const Composed = adopt({
  toggleCartHidden: ({ render }) => (
    <Mutation mutation={TOGGLE_CART_HIDDEN}>{render}</Mutation>
  ),
  localState: ({ render }) => <Query query={GET_CART_ITEMS}>{render}</Query>,
});

const CartDropDownComposed = props => (
  <Composed>
    {({ toggleCartHidden, localState }) => {
      const { cartItems } = localState.data;

      return (
        <CartDropDown
          toggleCartHidden={toggleCartHidden}
          cartItems={cartItems}
          {...props}
        />
      );
    }}
  </Composed>
);

export default withRouter(CartDropDownComposed);
