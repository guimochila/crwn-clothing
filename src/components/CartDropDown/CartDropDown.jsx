import React from 'react';

import CartItem from '../CartItem';
import {
  CartDropDownContainer,
  CartItemsContainer,
  CartDropDownButton,
  EmptyMessageContainer,
} from './CartDropDown.styles';

function CartDropDown({ cartItems, history, toggleCartHidden }) {
  return (
    <CartDropDownContainer>
      <CartItemsContainer>
        {cartItems.length ? (
          cartItems.map(cartItem => (
            <CartItem key={cartItem.id} item={cartItem} />
          ))
        ) : (
          <EmptyMessageContainer>Your cart is empty</EmptyMessageContainer>
        )}
      </CartItemsContainer>
      <CartDropDownButton
        onClick={() => {
          history.push('/checkout');
          toggleCartHidden();
        }}
      >
        GO TO CHECKOUT
      </CartDropDownButton>
    </CartDropDownContainer>
  );
}

export default CartDropDown;
