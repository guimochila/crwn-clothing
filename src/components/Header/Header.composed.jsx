import React from 'react';
import { Query } from 'react-apollo';

import Header from './Header';
import { GET_CART_HIDDEN } from '../../graphql/queries';

const HeaderComposed = () => (
  <Query query={GET_CART_HIDDEN}>
    {({ data: { cartHidden } }) => <Header hidden={cartHidden} />}
  </Query>
);

export default HeaderComposed;
