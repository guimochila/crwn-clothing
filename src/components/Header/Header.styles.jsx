import styled from 'styled-components';
import { Link } from 'react-router-dom';

export const HeaderContainer = styled.div`
  height: 70px;
  width: 100%;
  display: flex;
  justify-content: space-between;
  margin-bottom: 25px;
`;

export const LogoContainer = styled(Link)`
  height: 100%;
  width: 70px;
  padding: 25px;
`;

export const OptionsContainer = styled.div`
  width: 50%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: flex-end;
`;

export const OptionLink = styled(Link)`
  padding: 10px 15px;
  cursor: pointer;
`;

export const HeaderWelcome = styled.span`
  margin-left: 15px;
`;

export const ProfileImageContainer = styled.div`
  width: 40px;
  height: 40px;
  padding: 5px;
`;

export const ProfileImage = styled.span`
  background-size: 32px 32px;
  border-radius: 50%;
  display: block;
  margin: 0;
  overflow: hidden;
  height: 32px;
  width: 32px;

  &::before {
    display: inline-block;
    content: url(${({ photoURL }) => photoURL});
    transform: scale(0.5);
    transform-origin: left 0;
  }

  &:hover {
    box-shadow: 0 1px 0 rgba(0, 0, 0, 0.15);
  }
`;
