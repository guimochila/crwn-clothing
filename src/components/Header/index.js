import HeaderComposed from './Header.composed';

export default HeaderComposed;
