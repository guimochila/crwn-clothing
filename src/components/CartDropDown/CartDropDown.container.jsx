import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';

import CartDropDown from './CartDropDown';
import { selectCartItems } from '../../store/cart/cart.selectors';

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems,
});

const CartDropDownContainer = compose(
  withRouter,
  connect(mapStateToProps),
)(CartDropDown);

export default CartDropDownContainer;
