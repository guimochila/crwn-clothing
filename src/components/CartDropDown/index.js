import CartDropDown from './CartDropDown';
import CartDropDownContainer from './CartDropDown.container';

export { CartDropDown, CartDropDownContainer };
