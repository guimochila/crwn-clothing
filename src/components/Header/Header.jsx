import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { ReactComponent as Logo } from '../../assets/crown.svg';
import { selectCurrentUser } from '../../store/user/user.selectors';
import { selectCartHidden } from '../../store/cart/cart.selectors';
import { signOutStart } from '../../store/user/user.actions';
import CartIcon from '../CartIcon';
import { CartDropDownContainer } from '../CartDropDown';
import {
  HeaderContainer,
  LogoContainer,
  OptionsContainer,
  OptionLink,
  HeaderWelcome,
  ProfileImageContainer,
  ProfileImage,
} from './Header.styles';

function Header({ currentUser, hidden, signOutStart }) {
  const { displayName, photoURL } = currentUser || {};

  const firstName = displayName ? displayName.split(' ')[0] : null;

  return (
    <HeaderContainer>
      <LogoContainer to="/">
        <Logo className="logo" />
      </LogoContainer>
      <OptionsContainer>
        <OptionLink to="/shop">SHOP</OptionLink>
        <OptionLink to="/shop">CONTACT</OptionLink>
        {currentUser ? (
          <OptionLink as="div" onClick={signOutStart}>
            SIGN OUT
          </OptionLink>
        ) : (
          <OptionLink to="/signin">SIGN IN</OptionLink>
        )}
        <CartIcon />
        {firstName && <HeaderWelcome>Welcome {firstName}</HeaderWelcome>}
        <ProfileImageContainer>
          <ProfileImage photoURL={photoURL} />
        </ProfileImageContainer>
      </OptionsContainer>
      {hidden ? null : <CartDropDownContainer />}
    </HeaderContainer>
  );
}

const mapStateToProps = createStructuredSelector({
  currentUser: selectCurrentUser,
  hidden: selectCartHidden,
});

const mapDispatchToProps = dispatch => ({
  signOutStart: () => dispatch(signOutStart()),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Header);
