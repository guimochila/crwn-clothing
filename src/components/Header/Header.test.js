import React from 'react';
import { shallow } from 'enzyme';

import { Header } from './Header';
import { CartDropDownContainer } from '../CartDropDown';

describe('Header component', () => {
  let wrapper;
  let mockSignOutStart = jest.fn();
  const currentUser = {
    displayName: 'Anonymous Test',
    photoUrl: 'https://imgsrc.com/photo.jpg',
    uid: '123',
  };

  beforeEach(() => {
    const mockProps = {
      currentUser,
      signOutStart: mockSignOutStart,
      hidden: true,
    };

    wrapper = shallow(<Header {...mockProps} />);
  });

  test('should render Header component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should render sign out Link if currentUser is present', () => {
    expect(
      wrapper
        .find('OptionLink')
        .at(2)
        .text(),
    ).toBe('SIGN OUT');
  });

  test('should call signOutStart method when link is clicked', () => {
    wrapper
      .find('OptionLink')
      .at(2)
      .simulate('click');
    expect(mockSignOutStart).toHaveBeenCalled();
  });

  test('should render sign in link if currentUser is null', () => {
    const mockProps = {
      hidden: true,
      currentUser: null,
      signOutStart: mockSignOutStart,
    };
    const wrapper = shallow(<Header {...mockProps} />);

    expect(
      wrapper
        .find('OptionLink')
        .at(2)
        .text(),
    ).toBe('SIGN IN');
  });

  test('it should not render CartDropDown if hidden prop is true', () => {
    expect(wrapper.exists(CartDropDownContainer)).toBeFalsy();
  });

  test('it should render CartDropDown if hidden prop is false', () => {
    const mockProps = {
      hidden: false,
      currentUser: null,
      signOutStart: mockSignOutStart,
    };

    const wrapper = shallow(<Header {...mockProps} />);

    expect(wrapper.exists(CartDropDownContainer)).toBeTruthy();
  });
});
