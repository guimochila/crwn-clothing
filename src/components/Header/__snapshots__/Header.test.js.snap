// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render Header component 1`] = `
<styled.div>
  <Styled(Link)
    to="/"
  >
    <ForwardRef(SvgCrown)
      className="logo"
    />
  </Styled(Link)>
  <styled.div>
    <OptionLink
      to="/shop"
    >
      SHOP
    </OptionLink>
    <OptionLink
      to="/shop"
    >
      CONTACT
    </OptionLink>
    <OptionLink
      as="div"
      onClick={[MockFunction]}
    >
      SIGN OUT
    </OptionLink>
    <Connect(CartIcon) />
    <styled.span>
      Welcome 
      Anonymous
    </styled.span>
    <styled.div>
      <styled.span />
    </styled.div>
  </styled.div>
</styled.div>
`;
